appinventor/components/src/com/google/appinventor/components/runtime/FusiontablesControl.java

```diff
@@ -404,16 +407,50 @@ public void ForgetLogin() {
     OAuth2Helper.resetAccountCredential(activity);
   }
 
+  //InsertRow receives a String or a list as a value
   @SimpleFunction(
     description="Inserts a row into the specified fusion table. The tableId field is the id of the" +
       "fusion table. The columns is a comma-separated list of the columns to insert values into. The" +
       " values field specifies what values to insert into each column.")
-  public void InsertRow(String tableId, String columns, String values) {
-    query = "INSERT INTO " + tableId + " (" + columns + ")" + " VALUES " + "(" + values + ")";
+  public void InsertRow(String tableId, String columns, Object values) {
+    String reformattedValues = "";
+    if(values != null && values instanceof String || values != null && values instanceof gnu.lists.FString) {
+      Log.i("FusionTablesInsertRow", "String:"+values);
+      String valuesString = values.toString();
+      //Single quotes are escape chars and must be replaced with two single quotes (Natalie)
+      if(valuesString.startsWith("\'")){
+        reformattedValues = valuesString;
+      }
+      else if(valuesString.startsWith("\"")) {
+        String quotifiedValues = valuesString.replace("\'", "\'\'");
+        reformattedValues = quotifiedValues.replace("\"", "\'");
+      }
+      else{
+        //Not sure if we want to have a specific error message for this case.
+        //The code below is also probably an incorrect implementation of error messages (Natalie)
+        /*form.dispatchErrorOccurredEventDialog(this, "SendQueryValuesFormat",
+            ErrorMessages.FUSION_TABLES_QUERY_ERROR, query, "Values must either be in a list or be a" +
+                "String formatted as such: \'First\',\'Second\',\'Third\' Please note that single quotes" +
+                "in text must be inputted as two single quotes in a row: \'\'");
+        */
+        return;
+      }
+    }
+    else if(values != null && values instanceof YailList) {
+      Log.i("FusionTablesInsertRow", "YailList:"+values);
+      YailList valuesList = (YailList)values;
+      reformattedValues = "\'" + valuesList.getString(0).replace("\'", "\'\'") + "\'";
+      for (int i = 1; i < valuesList.size(); i++) {
+        reformattedValues += ",\'" + valuesList.getString(i).replace("\'", "\'\'") + "\'";
+      }
+    }
+    else {
+      Log.i("FusionTablesInsertRow", "Not String or YailList:"+values+" of class:"+values.getClass());
```

Comment:
Logging is okay, but to the user nothing will appear to be happening unless they view the log via `adb logcat`. We should report an error back to the blocks code so that users are made aware of the problem.

Also, while this should never happen in practice, there is no check whether `values` is null before attempting a call to `values.getClass()`, which would result in a NullPointerException.

Lastly, there should be a return here, otherwise reformattedValues will be the empty string and the query will be constructed and sent accordingly. This will likely result in an SQL execution error that will be sent back to the user.